### PR TITLE
MRG: Average with movement

### PR DESCRIPTION
**Pull request by Eric89GXL**
Adds movement compensation through averaging from [Taulu 2005] (http://lib.tkk.fi/Diss/2008/isbn9789512295654/article2.pdf). cc @staulu if you want to see how I implemented it in mne-python.

Requires `testing` dataset update.

@wronk do you have time to look at this one, too?

### CODE REVIEW

File: mne/__init__.py [hunk 1]
```diff
@@ -98,6 +98,7 @@
 from . import viz
 from . import decoding
 from . import realtime
+from . import testing
```
Comment (Eric89GXL) [inline mne/__init__.py]: @agramfort I think we should gradually move our testing-specific code here. For now I've only moved `assert_meg_snr`, which is now used by both `mne/preprocessing/tests/test_maxwell.py` and `mne/tests/test_epochs.py`. Should we name it `mne._testing` to make it private?

Comment (agramfort) [inline mne/__init__.py]: please keep all test code in tests folder. In could be in mne/tests/common.py

**Comment by agramfort** [comment]
how does it compare to mne_map_data?

how about we keep this feature in maxwell.py for now.

I am a bit worried to expose yet another method to the user. Epochs class is already a monster.

did you test it on real data already?

**Comment by Eric89GXL** [comment]
> how does it compare to mne_map_data?

Not sure, haven't looked at that yet. I need this for something I'm working on currently, but haven't done an extensive comparison. I guess I could add that one in, too, now that I think about it.

> how about we keep this feature in maxwell.py for now.

We can, but it will be something like `mne.preprocessing.maxwell.compensate_movement_epochs` or so, which is a bit long. Sound okay/preferable?

> I am a bit worried to expose yet another method to the user. Epochs class is already a monster.

Yeah agreed.

> did you test it on real data already?

Not much. I plan to use it heavily this week. The test has some simple validation comparing to movement-compensated data, which gave me some confidence it is correct. We could wait to merge until I've more heavily tested it, though.

File: mne/__init__.py (hunk 1 above)
Comment (Eric89GXL) [inline mne/__init__.py]: At least `numpy` (and probably other modules) have a `<name>.testing`, which is different from the actual tests, which often live in `<name>.tests`. Thus the motivation for the split. But if you'd prefer not to follow this convention, I can create the `common.py` file, and our imports will be `from mne.tests.common import ...`.

Comment (agramfort) [inline mne/__init__.py]: +1 for common.py

and +1 not to rush this merge before some extended test and validation.

didn't you already add the mne_map_data?

Comment (Eric89GXL) [inline mne/__init__.py]: Yeah, the equivalent has been in there for a while. It would be good to see how it compares to this version.

Comment (agramfort) [inline mne/__init__.py]: let me know what comparison gives.

Comment (Eric89GXL) [inline mne/__init__.py]: Will do. It's a high priority for me so I hope to get it done it time for release.

File: mne/preprocessing/maxwell.py [hunk 2]
```diff
@@ -446,6 +394,68 @@ def maxwell_filter(raw, origin='auto', int_order=8, ext_order=3,
...
+    n_bases = _get_n_moments([int_order, ext_order]).sum()
+    if n_bases > len(good_picks):
+        raise ValueError('Number of requested bases (%s) exceeds number of '
+                         'good sensors (%s)' % (str(n_bases), len(good_picks)))
+    recons = [ch for ch in info['bads']
+              if info['ch_names'].index(ch) in meg_picks]
+    if len(recons) > 0:
+        logger.info('    Bad MEG channels being reconstructed: %s' % recons)
+    else:
+        logger.info('    No bad MEG channels')
```
Comment (wronk) [inline mne/preprocessing/maxwell.py]: Any reason the above two sets of spaces aren't `\t`?

Comment (Eric89GXL) [inline mne/preprocessing/maxwell.py]: we just always use four spaces in mne-python

**Comment by wronk** [comment]
Could you can add a sentence or two to the `average_movements` docstring about use cases/what problems it's addressing? Other than that, LGTM

**Comment by Eric89GXL** [comment]
@wronk thanks for the comments, will address soon.

@agramfort I implemented the `map_data` method, but it's *really* slow. Like 10-20 seconds per head position slow. So I don't think it's feasible to use it. Also, Matti had a validation paper where he only used the Maxwell filtering method, so I think there's probably a good reason it wasn't tested as a viable alternative.

**Comment by jasmainak** [comment]
@Eric89GXL did you cache the calls for `map_data` using joblibs? That might make it faster.

**Comment by Eric89GXL** [comment]
I don't think it will help, the number of possible head positions is effectively infinite. I guess if you ran the same file over and over again, it would be faster on subsequent executions, but that doesn't seem too helpful.

**Comment by jasmainak** [comment]
hmm ... that's true. How long does it take in C?

**Comment by Eric89GXL** [comment]
Not sure. But we should be close to C speed because I put in some additional vectorization, so now the loop is only over sensors (e.g., 306) instead of sensors x sensors, so the interpretation overhead should be small.

File: mne/forward/_lead_dots.py [hunk 3]
```diff
@@ -185,8 +186,12 @@ def _comp_sums_meg(beta, ctheta, lut_fun, n_fact, volume_integral):
...
 # SPHERE DOTS
 
-def _fast_sphere_dot_r0(r, rr1, rr2, lr1, lr2, cosmags1, cosmags2,
-                        w1, w2, volume_integral, lut, n_fact, ch_type):
+_meg_const = 4e-14 * np.pi  # This is \mu_0^2/4\pi
+_eeg_const = 1.0 / (4.0 * np.pi)
+
+
+def _fast_sphere_dot_r0(r, rr1_orig, rr2s, lr1, lr2s, cosmags1, cosmags2s,
+                        w1, w2s, volume_integral, lut, n_fact, ch_type):
```
Comment (Eric89GXL) [inline mne/forward/_lead_dots.py]: @jasmainak see changes to this function for improved vectorization of computations

**Comment by Eric89GXL** [comment]
Ready for review/merge from my end. I've validated this on about ~30 datasets created with the raw data simulator with movement and it seems to work as well as MaxFilter's movement compensation algorithm.

File: mne/tests/common.py [hunk 4]
```diff
@@ -0,0 +1,48 @@
+# Authors: Eric Larson <larson.eric.d@gmail.com>
+#
+# License: BSD (3-clause)
+
+import numpy as np
+from numpy.testing import assert_allclose
+
+from .. import pick_types, Evoked
+from ..io import _BaseRaw
```
Comment (agramfort) [inline mne/tests/common.py]: don't use relative imports in tests

**Comment by agramfort** [comment]
besides +1 for merge

although I did not get everything :) I hope you don't get hit by a bus...

File: mne/tests/common.py (hunk 4 above)
Comment (Eric89GXL) [inline mne/tests/common.py]: Hmm... this isn't really a *test*, though, it's a set of functions used *by tests* (but in theory could be used other places).

Comment (agramfort) [inline mne/tests/common.py]: ok you win....

**Comment by agramfort** [comment]
thanks heaps @Eric89GXL and @staulu !

**Comment by Eric89GXL** [comment]
Thanks for the look. 0.11 is approaching...

**Pull request merged**